jwst/background/background_sub.py

```diff
@@ -91,3 +109,191 @@ def average_background(bkg_list):
+        got_catalog = True
+    else:
+        log.warning("No source_catalog found in input.meta.")
+        got_catalog = False
+
+    # If there are NaNs, we have to replace them with something harmless.
+    bkg_ref = no_NaN(bkg_ref)
+
+    # Create a mask from the source catalog, True where there are no sources,
+    # i.e. in regions we can use as background.
+    if got_catalog:
+        bkg_mask = mask_from_source_cat(input_model.data.shape, source_catalog)
+    else:
+        bkg_mask = np.ones(input_model.data.shape, dtype=np.bool)
+
+    # Compute the mean values of science image and background reference
+    # image, including only regions where there are no identified sources.
+    # Exclude pixel values in the lower and upper 25% of the histogram.
+    lowlim = 25.
+    highlim = 75.
+    sci_mean = robust_mean(input_model.data[bkg_mask],
+                           lowlim=lowlim, highlim=highlim)
+    bkg_mean = robust_mean(bkg_ref.data[bkg_mask],
+                           lowlim=lowlim, highlim=highlim)
+
+    log.debug("mean of [{}, {}] percentile science data = {}"
+              .format(lowlim, highlim, sci_mean))
+    log.debug("mean of [{}, {}] percentile background image = {}"
+              .format(lowlim, highlim, bkg_mean))
+
+    result = input_model.copy()
+    if bkg_mean != 0.:
+        subtract_this = (sci_mean / bkg_mean) * bkg_ref.data
+        result.data = input_model.data - subtract_this
+        log.debug("Average of values subtracted = {}"
+                  .format(subtract_this.mean(dtype=np.float)))
+    else:
+        log.warning("Background file has zero mean; "
+                    "nothing will be subtracted.")
+    result.dq = np.bitwise_or(input_model.dq, bkg_ref.dq)
+    return result
+
+
+def no_NaN(model, fill_value=0.):
+    """Replace NaNs with a harmless value.
+
+    Parameters
+    ----------
+    model: JWST data model
+        Reference file model.
+
+    fill_value: float
+        NaNs will be replaced with this value.
+
+    Returns
+    -------
+    result: JWST data model
+        Reference file model without NaNs in data array.
+    """
+
+    mask = np.isnan(model.data)
+    if mask.sum(dtype=np.intp) == 0:
+        return model
+    else:
+        temp = model.copy()
+        temp.data[mask] = fill_value
+        return temp
+
+
+def mask_from_source_cat(shape, source_catalog):
+    """Create a mask that is False within bounding boxes of sources.
+
+    Parameters
+    ----------
+    shape: tuple
+        The boolean mask will be created with this shape.
+
+    source_catalog: str
+        Name of the source catalog (a .ecsv file).
+
+    Returns
+    -------
+    bkg_mask: ndarray
+        Boolean mask:  True for background, False for pixels that are
+        inside at least one of the source regions defined in the source
+        catalog.
+    """
+
+    # This was mostly copied from get_object_info() in assign_wcs/util.py
```

Comment:
You can use create_grism_bbox directly here, that function calls get_object_info and checks that the required catalog fields are present. It returns a list of GrismObjects containing information about all the sources as well as the appropriate bounding boxes around the trace for each spectral order for each object.